### PR TITLE
Refactor Chatbot to use Astrapy instead of JSON API

### PR DESCRIPTION
**Pull request by ElenaKusevska**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,8 @@
+ASTRA_DB_APPLICATION_TOKEN=
+ASTRA_DB_API_ENDPOINT=
+ASTRA_DB_KEYSPACE=
```
Comment (jgillenwater) [inline .env.example]: Why did we switch this to KEYSPACE? IIUC, we are using keyspace for CQL/drivers, but namespace for everything else.

File: README.md [hunk 2]
```diff
@@ -8,15 +8,24 @@
...
 ## Setting up your database and seeding with data
 1. Navigate to your IDE, set up the following environment variables:
 
-- ASTRA_DB_NAMESPACE=existing Astra Namespace in a vector enabled DB
-- OPENAI_API_KEY=api key for OPENAI
-- ASTRA_DB_ID=Astra DB database id
-- ASTRA_DB_REGION=Astra DB database region
 - ASTRA_DB_APPLICATION_TOKEN=Generate app token for Astra database
+- ASTRA_DB_KEYSPACE=existing Astra Namespace/keyspace in a vector enabled DB
+- ASTRA_DB_API_ENDPOINT="https://ASTRA_DB_ID-ASTRA_REGIN.apps.astra.datastax.com
```
Comment (jgillenwater) [inline README.md]:
```suggestion
- ASTRA_DB_API_ENDPOINT="https://ASTRA_DB_ID-ASTRA_REGION.apps.astra.datastax.com"
```

File: README.md [hunk 3]
```diff
@@ -8,15 +8,24 @@
...
+- ASTRA_DB_COLLECTION_NAME=Name of collection/table to be created in Astra database
+- VECTOR_DIMENSION=Collection to be created with number of vector dimensions
+- SCRAPED_FILE=Scrapped data file location
```
Comment (jgillenwater) [inline README.md]:
```suggestion
- SCRAPED_FILE=Scraped data file location
```

File: api/API_README.md [hunk 4]
```diff
@@ -10,7 +10,16 @@ git clone https://github.com/Anant/astra-chatbot-react-python.git
...
 
-Input all of you authentication credentials into utils/local_creds.py. This file requries your Astra app token, database id, and region. It also needs the keyspace name and collection name for your vector store. Lastly it requries your OpenAI api key.
+Input all authentication credentials into .env file. This file requires below Astra information.
+
+    - ASTRA_DB_APPLICATION_TOKEN=Generate app token for Astra database
+    - ASTRA_DB_KEYSPACE=existing Astra Namespace/keyspace in a vector enabled DB
+    - ASTRA_DB_API_ENDPOINT="https://ASTRA_DB_ID-ASTRA_REGIN.apps.astra.datastax.com
+    - ASTRA_DB_COLLECTION_NAME=Name of collection/table to be created in Astra database
+    - VECTOR_DIMENSION=Collection to be created with number of vector dimensions
+    - SCRAPED_FILE=Scrapped data file location
```
Comment (jgillenwater) [inline api/API_README.md]:
```suggestion
    - SCRAPED_FILE=Scraped data file location
```

File: README.md [hunk 5]
```diff
@@ -8,15 +8,24 @@
...
 
 ## Setting up your database and seeding with data
 1. Navigate to your IDE, set up the following environment variables:
 
-- ASTRA_DB_NAMESPACE=existing Astra Namespace in a vector enabled DB
-- OPENAI_API_KEY=api key for OPENAI
-- ASTRA_DB_ID=Astra DB database id
-- ASTRA_DB_REGION=Astra DB database region
 - ASTRA_DB_APPLICATION_TOKEN=Generate app token for Astra database
+- ASTRA_DB_KEYSPACE=existing Astra Namespace/keyspace in a vector enabled DB
```
Comment (jgillenwater) [inline README.md]:
```suggestion
- ASTRA_DB_NAMESPACE=existing Astra DB namespace in vector database
```

File: api/API_README.md [hunk 6]
```diff
@@ -10,7 +10,16 @@ git clone https://github.com/Anant/astra-chatbot-react-python.git
 pip3 install -r requirements.txt
 ```
 
-Input all of you authentication credentials into utils/local_creds.py. This file requries your Astra app token, database id, and region. It also needs the keyspace name and collection name for your vector store. Lastly it requries your OpenAI api key.
+Input all authentication credentials into .env file. This file requires below Astra information.
+
+    - ASTRA_DB_APPLICATION_TOKEN=Generate app token for Astra database
+    - ASTRA_DB_KEYSPACE=existing Astra Namespace/keyspace in a vector enabled DB
```
Comment (jgillenwater) [inline api/API_README.md]:
```suggestion
    - ASTRA_DB_NAMESPACE=existing Astra DB namespace in vector databaseB
```

File: api/chatbot_utils.py [hunk 7]
```diff
@@ -1,48 +1,59 @@
...
+import os
+from dotenv import load_dotenv
+from astrapy.db import AstraDBCollection
+
+load_dotenv()
+
+# Grab the Astra token and api endpoint from the environment
+token = os.getenv("ASTRA_DB_APPLICATION_TOKEN")
+api_endpoint = os.getenv("ASTRA_DB_API_ENDPOINT")
+keyspace = os.getenv("ASTRA_DB_KEYSPACE")
```
Comment (jgillenwater) [inline api/chatbot_utils.py]:
```suggestion
keyspace = os.getenv("ASTRA_DB_NAMESPACE")
```

File: populate_db/create_collection.py [hunk 8]
```diff
@@ -1,25 +1,36 @@
...
-payload = json.dumps({"createCollection": {
-    "name": "chat",
-    "options" : {
-        "vector" : {
-            "size" : 1536,
-            "function" : "cosine"}}}})
+# Grab the Astra token and api endpoint from the environment
+token = os.getenv("ASTRA_DB_APPLICATION_TOKEN")
+api_endpoint = os.getenv("ASTRA_DB_API_ENDPOINT")
+keyspace = os.getenv("ASTRA_DB_KEYSPACE")
```
Comment (jgillenwater) [inline populate_db/create_collection.py]:
```suggestion
keyspace = os.getenv("ASTRA_DB_NAMESPACE")
```

File: populate_db/load_data.py [hunk 9]
```diff
@@ -1,23 +1,37 @@
...
-embeddings = OpenAIEmbeddings(openai_api_key=OPENAI_API_KEY)
 
-request_url = f"https://{ASTRA_DB_ID}-{ASTRA_DB_REGION}.apps.astra.datastax.com/api/json/v1/{ASTRA_DB_NAMESPACE}/chat"
-request_headers = { 'x-cassandra-token': ASTRA_DB_APPLICATION_TOKEN,  'Content-Type': 'application/json'}
+load_dotenv()
+
+# Grab the Astra token and api endpoint from the environment
+token = os.getenv("ASTRA_DB_APPLICATION_TOKEN")
+api_endpoint = os.getenv("ASTRA_DB_API_ENDPOINT")
+keyspace = os.getenv("ASTRA_DB_KEYSPACE")
```
Comment (jgillenwater) [inline populate_db/load_data.py]:
```suggestion
keyspace = os.getenv("ASTRA_DB_NAMESPACE")
```

File: .env.example [hunk 10]
```diff
@@ -0,0 +1,8 @@
...
+ASTRA_DB_COLLECTION_NAME=
```
Comment (jgillenwater) [inline .env.example]:
```suggestion
ASTRA_DB_COLLECTION=
```

File: README.md [hunk 11]
```diff
@@ -8,15 +8,24 @@
...
+- ASTRA_DB_API_ENDPOINT="https://ASTRA_DB_ID-ASTRA_REGIN.apps.astra.datastax.com
+- ASTRA_DB_COLLECTION_NAME=Name of collection/table to be created in Astra database
```
Comment (jgillenwater) [inline README.md]:
```suggestion
- ASTRA_DB_COLLECTION=Name of collection/table to be created in Astra database
```

File: api/API_README.md [hunk 12]
```diff
@@ -10,7 +10,16 @@ git clone https://github.com/Anant/astra-chatbot-react-python.git
...
+    - ASTRA_DB_API_ENDPOINT="https://ASTRA_DB_ID-ASTRA_REGIN.apps.astra.datastax.com
+    - ASTRA_DB_COLLECTION_NAME=Name of collection/table to be created in Astra database
```
Comment (jgillenwater) [inline api/API_README.md]:
```suggestion
    - ASTRA_DB_COLLECTION=Name of collection/table to be created in Astra database
```

File: api/chatbot_utils.py [hunk 13]
```diff
@@ -1,48 +1,59 @@
...
+openai_api_key = os.getenv("OPENAI_API_KEY")
+collection_name = os.getenv("ASTRA_DB_COLLECTION_NAME")
```
Comment (jgillenwater) [inline api/chatbot_utils.py]:
```suggestion
collection_name = os.getenv("ASTRA_DB_COLLECTION")
```

File: populate_db/create_collection.py [hunk 14]
```diff
@@ -1,25 +1,36 @@
...
+collection_name = os.getenv("ASTRA_DB_COLLECTION_NAME")
```
Comment (jgillenwater) [inline populate_db/create_collection.py]:
```suggestion
collection_name = os.getenv("ASTRA_DB_COLLECTION")
```

File: populate_db/load_data.py [hunk 15]
```diff
@@ -1,23 +1,37 @@
...
+openai_api_key = os.getenv("OPENAI_API_KEY")
+collection_name = os.getenv("ASTRA_DB_COLLECTION_NAME")
```
Comment (jgillenwater) [inline populate_db/load_data.py]:
```suggestion
collection_name = os.getenv("ASTRA_DB_COLLECTION")
```

**Review by jgillenwater — CHANGES_REQUESTED** [review]
I made some suggestions so that we are using consistent env vars across tutorials. I also fixed a couple of minor typos.

File: .env.example (hunk 1 above)
Comment (hc33brackles) [reply .env.example]: @jgillenwater - this did get changed to namespace everywhere else - strongly think we should change it here as well

File: .env.example [hunk 16]
```diff
@@ -1,7 +1,7 @@
 ASTRA_DB_APPLICATION_TOKEN=
 ASTRA_DB_API_ENDPOINT=
 ASTRA_DB_KEYSPACE=
```
Comment (jgillenwater) [inline .env.example]:
```suggestion
ASTRA_DB_NAMESPACE=
```

File: .env.example (hunk 1 above)
Comment (jgillenwater) [reply .env.example]: Yep. Just submitted the same suggestion.

**Pull request merged**